### PR TITLE
WIP get rid of merging overlays

### PR DESCRIPTION
**Pull request by kunmingg**
**Which issue is resolved by this Pull Request:**
Related https://github.com/kubeflow/manifests/issues/774

**Description of your changes:**
* flatten overlays; 
* create kustomization file as signal entry point for `apply`.
* make kustomize package deployable on its own (through `kustomize` CLI)

For example for `jupyter-web-app`, before the change we need to merge overlays `application` and `istio` with `kfctl` logic since user need both of them:
```
bases:
...
overlays:
- application
- istio
```
However merge overlays is not supported by `kustomize`, maintain this feature in `kfctl` is expensive and will block user deploying our pkg with `kustomize` CLI directly.

After this PR merge, user can just `kustomize build base` directly:
```
bases:
- application
- base
- istio
- kustomization.yaml
```

`overlays` are still supported but only one of them will be used during deployment: `kustomize build overlays/name`

update test generation to allow nested `kustomization.yaml` under `base`

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/manifests/873)

### CODE REVIEW

**Comment by kunmingg** [comment]
@jlewi 
Can you review the changes on `jupyter-web-app`? 
If looks good will apply same change to other components.

**Comment by jlewi** [comment]
@kunmingg This is really elegant. So If I understand correctly if someone just wanted
to install the jupyter web app without the application CR or ISTIO they could do

```
kustomize build jupyter-web-app/bases 
```

If they want to install the jupyter web-app with the application resource they would do

```
kustomize build jupyter-web-app
```

If we wanted to define some overlays for GCP; e.g. if we wanted to override the jupyter-web-app configmap we would define an overlay whose base was jupyter-web-app?

How about including a markdown doc in this PR that describes the layout and changes? e.g. "manifests_v2.md"? We could mark it as a proposal and then discuss and circulate on the mailing list.

Related to that it might be useful to demonstrate the use of composition of applications. So define a Kubeflow manifest that combines the TFJob and JupyterWebApp manifests? This would essentially eliminate the use of KFDef to list all the KF applications. 

Would it be useful to put these new applications in a different subdirectory? e.g.
experimental/jupyter
experimental/tfjob

This would potentially allow us to check in the code while we experiment and socialize the idea?

**Comment by jlewi** [comment]
@kunmingg WDYT the next steps are here?

I think I've changed course a bit from doing this in experimental. I think I'm leaning towards doing this in a more incremental way. In particular, I think we should consider moving forward refactoring each application manifest and updating the KFDefs to remove the dependency on applying multiple overlays. 

/cc @kimwnasptd

**Comment by kunmingg** [comment]
@jlewi Sounds good, let me update this PR with docs explaining new behaviors.

**Comment by jlewi** [comment]
Actually if we look it at what kfctl does it just ends up generating the following kustomization file

```
apiVersion: kustomize.config.k8s.io/v1beta1
bases:
- base
- overlays/istio
commonLabels:
  app.kubernetes.io/component: jupyter-web-app
  app.kubernetes.io/instance: jupyter-web-app-v1.0.0
  app.kubernetes.io/managed-by: kfctl
  app.kubernetes.io/name: jupyter-web-app
  app.kubernetes.io/part-of: kubeflow
  app.kubernetes.io/version: v1.0.0
kind: Kustomization
namespace: kubeflow
resources:
- overlays/application/application.yaml
```

I can't think of any reason to autogenerate that file as opposed to just checking it in. All though bases is deprecated so we could just change it to

```
apiVersion: kustomize.config.k8s.io/v1beta1
commonLabels:
  app.kubernetes.io/component: jupyter-web-app
  app.kubernetes.io/instance: jupyter-web-app-v1.0.0
  app.kubernetes.io/managed-by: kfctl
  app.kubernetes.io/name: jupyter-web-app
  app.kubernetes.io/part-of: kubeflow
  app.kubernetes.io/version: v1.0.0
kind: Kustomization
namespace: kubeflow
resources:
- base
- overlays/istio
- overlays/application/application.yaml
```

I think we might need to add the configurations to this in order to handle substituting namespace and cluster prefix into the virtual service.

I don't think we really want the files
overlays/istio/kustomization.yaml
overlays/application/kustomization.yaml

That implies you can run kustomize build in that directory but that doesn't really make sense. The resources defined there only make sense as additions to the base package.

That said it might be worth keeping them for some sort of backwards compatibility.

**Comment by jlewi** [comment]
@kunmingg I was doing a bunch of experiments; see https://github.com/kubeflow/manifests/pull/962

TL;DR lets chat before you invest more time in this PR.
